### PR TITLE
Wallectconnect integration with web

### PR DESCRIPTION
**Pull request by ratik21**
## Proposed Changes

+ add `WallectConnectSession` class to create new/use an existing wallet connect session. 
+ move `getSuggestedParams`, `mkTxParams` from runtime to web.

## Potential followups
+ Update existing dapp/create new dapp using walletconnect.
+ Handle case of multiple signers in group transaction (txns signed by multiple users in single group tx)

### CODE REVIEW

File: packages/web/src/lib/api.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ */
+export async function mkTxParams (
+  algocl: Algodv2, userParams: TxParams, s?: SuggestedParams): Promise<SuggestedParams> {
+  if (s === undefined) { s = await getSuggestedParams(algocl); }
+
+  if (userParams.flatFee === undefined) {
+    if (userParams.totalFee !== undefined) s.flatFee = true;
+    else s.flatFee = false;
+  }
+  s.fee = userParams.totalFee ?? userParams.feePerByte ?? ALGORAND_MIN_TX_FEE;
```
Comment (robert-zaremba) [inline packages/web/src/lib/api.ts:47]: I think we shouldn't set `ALGORAND_MIN_TX_FEE` if both `userParams.totalFee` and `userParams.feePerByte` are unknown.

**Review by robert-zaremba — COMMENTED** [review]
Thanks for updating the comments. We need to add changelog entry.

File: packages/web/src/types.ts [hunk 2]
```diff
@@ -291,3 +293,44 @@ export function isSDKTransactionAndSign (object: unknown): object is Transaction
   const res = isSDKTransaction((object as TransactionAndSign).transaction);
   return Object.prototype.hasOwnProperty.call(object, "sign") && res;
 }
+
+/* Wallet Connect types */
+
+export enum ChainType {
+  MainNet = "MainNet",
+  TestNet = "TestNet"
```
[the remark this answers is not in the document]
Comment (ratik21) [reply packages/web/src/types.ts:301]: yes, in code we can (we just need extra info of url/host, and use our infrastructure config as default).

But when i used WalletConnect from my Algorand Wallet, in node settings i was only able to select MainNet/TestNet. So not sure if betanet is supported

File: packages/web/src/types.ts [hunk 3]
```diff
@@ -291,3 +293,44 @@ export function isSDKTransactionAndSign (object: unknown): object is Transaction
...
+}
+
+export interface SessionConnectResponse {
+  peerId: string
+  peerMeta?: IClientMeta
+  accounts: string[]
+}
```
[the remark this answers is not in the document]
Comment (ratik21) [reply packages/web/src/types.ts:308]: ok, renaming to `wcPeerID`, `wcAccounts` ..

[the remark this answers is not in the document]
Comment (ratik21) [reply packages/web/src/types.ts:308]: But we can't rename this - as this is a response from the `onConnect` subscription api by wallet connect. Renaming `accounts` in our session class

File: packages/web/src/algo-signer-types.ts [hunk 4]
```diff
@@ -29,17 +29,51 @@ export interface MultisigTransaction {
 }
 
 export interface WalletMultisigMetadata {
-  readonly version: number
-  readonly threshold: number
-  readonly addrs: string[]
+  /**
+   * Multisig version.
+   */
```
[the remark this answers is not in the document]
Comment (ratik21) [reply packages/web/src/algo-signer-types.ts:34]: got it

File: packages/web/src/lib/api.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * Returns blockchain transaction suggested parameters (firstRound, lastRound, fee..)
+ * @param algocl an Algorand client, instance of Algodv2, used to communicate with a blockchain node.
+ */
+export async function getSuggestedParams (algocl: Algodv2): Promise<SuggestedParams> {
+  const params = await algocl.getTransactionParams().do();
+  // Private chains may have an issue with firstRound
+  if (params.firstRound === 0) {
+    throw new Error("Suggested params returned 0 as firstRound. Ensure that your node progresses.");
+    // params.firstRound = 1
```
[the remark this answers is not in the document]
Comment (ratik21) [reply packages/web/src/lib/api.ts:28]: yes, removed

File: packages/web/src/types.ts [hunk 6]
```diff
@@ -291,3 +293,44 @@ export function isSDKTransactionAndSign (object: unknown): object is Transaction
...
+  /**
+     * Optional message explaining the reason of the group of
+     * transactions.
+     */
+  message?: string
+
+  // other options may be present, but are not standard
+}
+
+export type SignTxnParams = [WalletTransaction[], SignTxnOpts?];
```
[the remark this answers is not in the document]
Comment (ratik21) [reply packages/web/src/types.ts:328]: agree, but this is how the transaction should be before sending `formatJsonRpcRequest` to sign transaction. Hence the type. Check [this](https://github.com/scale-it/algo-builder/pull/525/files/047c047f5047cf115c39386ca54d0ab9fe82742f#diff-2d2f4883af18a542ace818cf8798b3d5d1ebcbdc08b6963a7692bc8d6ca4e2d0R146) line

File: packages/web/src/lib/api.ts (hunk 1 above)
Comment (ratik21) [reply packages/web/src/lib/api.ts:46]: Could you please explain the reason (not setting fee as minimum, if not passed by user) ?

OR do you mean the network should set fee on it's own in this case?

Comment (robert-zaremba) [reply packages/web/src/lib/api.ts:45]: Yes, the JS SDK will do it.

**Review by robert-zaremba — APPROVED** [review]
Looks good. let's solve the comment about fees before merging.

File: packages/web/src/lib/api.ts (hunk 1 above)
Comment (ratik21) [reply packages/web/src/lib/api.ts:45]: ok

Comment (ratik21) [reply packages/web/src/lib/api.ts:45]: ah, we can't set the fee as `undefined`, it must be number. [https://github.com/algorand/js-algorand-sdk/blob/develop/src/transaction.ts#L476](https://github.com/algorand/js-algorand-sdk/blob/develop/src/transaction.ts#L476)

Comment (ratik21) [reply packages/web/src/lib/api.ts:45]: type [https://github.com/algorand/js-algorand-sdk/blob/develop/src/types/transactions/base.ts#L94](https://github.com/algorand/js-algorand-sdk/blob/develop/src/types/transactions/base.ts#L94)

**Pull request merged**